gradle-plugins/jaicp-deploy-plugin/src/main/kotlin/com/justai/jaicf/plugins/jaicpdeploy/JaicpDeployPlugin.kt

```diff
@@ -0,0 +1,57 @@
+package com.justai.jaicf.plugins.jaicpdeploy
+
+import com.github.jengelman.gradle.plugins.shadow.ShadowJavaPlugin
+import com.github.jengelman.gradle.plugins.shadow.ShadowPlugin
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import org.gradle.api.DefaultTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.tasks.TaskAction
+
+class JaicpDeployPlugin : Plugin<Project> {
+
+    override fun apply(project: Project): Unit = with(project) {
+        if (!isRootProject) {
+            logger.warn("JAICP deploy plugin should only be applied on a root project")
+            return
+        }
+
+        applyPluginSafely(JavaPlugin::class)
+        applyPluginSafely(ShadowPlugin::class)
+
+        tasks.register(JAICP_BUILD_TASK_NAME, JaicpBuild::class.java) {
+            it.finalizedBy(tasks.withType(ShadowJar::class.java).named(SHADOW_JAR_TASK_NAME))
+        }
+    }
+
+}
+
+open class JaicpBuild : DefaultTask() {
+    private val jarDestinationDir: String = project.prop(JAR_DESTINATION_DIR) ?: JAR_DESTINATION_DIR_DEFAULT
+    private val jarFileName: String = project.prop(JAR_FILE_NAME) ?: JAR_FILE_NAME_DEFAULT
+    var mainClassName: String? = null
+
+    @TaskAction
+    fun action() {
+        project.tasks.withType(ShadowJar::class.java).named(SHADOW_JAR_TASK_NAME) { jar ->
+            jar.destinationDirectory.set(project.file(jarDestinationDir))
+            jar.archiveFileName.set(jarFileName)
+
+            val mainClass = mainClassName ?: jar.manifest.attributes["Main-Class"]
+            mainClass?.let { jar.manifest.attributes["Main-Class"] = it }
+                ?: project.logger.warn("Main class name was not provided")
```

Comment:
Shouldn't we fail and throw exception when main class name was not provided?

Reply:
I think we shouldn't, because the Gradle jar task is not the only way to set the main class. A user can have some `dependsOn`, or `finalizedBy` tasks, that set proper main class.